Detectors/FIT/FV0/base/src/Geometry.cxx

```diff
@@ -7,117 +7,1183 @@
 // granted to it by virtue of its status as an Intergovernmental Organization
 // or submit itself to any jurisdiction.
-#include <iomanip>
-//#include <TVector3.h>
+
+/// \file   Geometry.cxx
+/// \brief  Implementation of FV0 geometry.
+///
+/// \author Maciej Slupecki, University of Jyvaskyla, Finland
+/// \author Andreas Molander, University of Helsinki, Finland
+
 #include "FV0Base/Geometry.h"
 
-#include <TGeoManager.h>
+#include <cmath>
+
+#include <FairLogger.h>
+
 #include <TGeoBBox.h>
-#include <TGeoTube.h>
 #include <TGeoCompositeShape.h>
+#include <TGeoCone.h>
+#include <TGeoManager.h>
+#include <TGeoMatrix.h>
 #include <TGeoMedium.h>
+#include <TGeoTube.h>
 #include <TGeoVolume.h>
-#include <TGeoMatrix.h>
-#include <FairLogger.h>
-#include <sstream>
 
 ClassImp(o2::fv0::Geometry);
 
 using namespace o2::fv0;
 
-Geometry::Geometry(EGeoType initType)
+Geometry::Geometry(EGeoType initType) : mGeometryType(initType)
 {
-  mGeometryType = initType;
-  initializeVectors();
-  initializeLuts();
-  buildGeometry();
+  initializeGeometry();
 }
 
-Geometry::Geometry(const Geometry& geom)
+Geometry::Geometry(const Geometry& geometry) : mGeometryType(geometry.mGeometryType), mLeftTransformation(nullptr), mRightTransformation(nullptr)
 {
-  this->mGeometryType = geom.mGeometryType;
+  this->mEnabledComponents = geometry.mEnabledComponents;
 }
 
-void Geometry::initializeVectors()
+Geometry::~Geometry()
 {
-  // RADII
-  // Index of mvrAvgScint is NOT linked directly to any ring number
-  mvrAvgScint.assign(sRingRadiiScint, sRingRadiiScint + sNumberOfRings + 1);
-
-  // Set real plastic radi (reduced by painting of 0.06 and separation gap of 0.08)
-  for (uint16_t ir = 1; ir < mvrAvgScint.size(); ir++) { // shift of indices to match index with ring number (starting from 0)
-    mvrMaxScint.push_back(mvrAvgScint.at(ir) - sDrSeparationScint);
+  if (mRightTransformation) {
+    delete mRightTransformation;
   }
-  for (uint16_t ir = 0; ir < mvrAvgScint.size() - 1; ir++) {
-    mvrMinScint.push_back(mvrAvgScint.at(ir) + sDrSeparationScint);
+  if (mLeftTransformation) {
+    delete mLeftTransformation;
   }
-  // Now indices of rMinScint and rMaxScint correspond to the same ring
+}
+
+const int Geometry::getCurrentCellId(const TVirtualMC* fMC) const
+{
+  int detectorHalfID = -1;
+  int sectorID = -1;
+  int ringID = -1;
+
+  fMC->CurrentVolOffID(2, detectorHalfID);
+  fMC->CurrentVolOffID(1, sectorID);
+  fMC->CurrentVolOffID(0, ringID);
+
+  sectorID += detectorHalfID * sNumberOfCellSectors;
 
-  // AZIMUTH
-  for (uint16_t isector = 0; isector < sBaseNumberOfSectors; isector++) {
-    float phiMin = sPhiMinScint + isector * sDphiScint + sGlobalPhiRotation;
-    std::stringstream ssNameRot;
-    ssNameRot << "rotPhiSector" << isector;
-    mvPhiRot.push_back(new TGeoRotation(ssNameRot.str().c_str(), phiMin, 0.0, 0.0));
-    mvPhiRot.at(mvPhiRot.size() - 1)->RegisterYourself();
+  LOG(DEBUG) << "FV0 Geometry::getCurrentCellId():";
```

Comment:
note that each ``LOG(DEBUG)`` will cost at an 'if' comparison (if not it debug mode). Since this method is called in every hit processing, you better wrap the log output to a single debug message.

Reply:
Done